pageserver/src/bin/layer_map_analyzer.rs

```diff
@@ -0,0 +1,212 @@
+            if let Some(prev) = prev_key {
+                if curr.to_i128() - prev.to_i128() >= MIN_HOLE_LENGTH {
+                    heap.push(Hole(prev..curr));
+                    if heap.len() > max_holes {
+                        heap.pop(); // remove smallest hole
+                    }
+                }
+            }
+            prev_key = Some(curr.next());
+            true
+        },
+    )?;
+    let mut holes = heap.into_vec();
+    holes.sort_by_key(|hole| hole.0.start);
+    Ok(holes)
+}
+
+fn main() -> Result<()> {
+    let args: Vec<String> = env::args().collect();
+    if args.len() < 2 {
+        println!("Usage: layer_map_analyzer PAGESERVER_DATA_DIR [MAX_HOLES]");
+        return Ok(());
+    }
+    let storage_path = PathBuf::from_str(&args[1])?;
+    let max_holes = if args.len() > 2 {
+        args[2].parse::<usize>().unwrap()
+    } else {
+        DEFAULT_MAX_HOLES
+    };
+    pageserver::virtual_file::init(10);
+    pageserver::page_cache::init(100);
+    let mut total_delta_layers = 0usize;
+    let mut total_image_layers = 0usize;
+    let mut total_excess_layers = 0usize;
+    for tenant in fs::read_dir(storage_path.join("tenants"))? {
+        let tenant = tenant?;
+        if !tenant.file_type()?.is_dir() {
+            continue;
+        }
+        for timeline in fs::read_dir(tenant.path().join("timelines"))? {
+            let timeline = timeline?;
+            if !timeline.file_type()?.is_dir() {
+                continue;
+            }
+            let mut layers = Vec::new();
+            let mut n_deltas = 0usize;
+            let mut n_excess_layers = 0usize;
+            let mut n_holes = 0usize;
+            for layer in fs::read_dir(timeline.path())? {
+                let layer = layer?;
+                if let Some(mut layer_file) =
+                    parse_filename(&layer.file_name().into_string().unwrap())
+                {
+                    if layer_file.is_delta {
+                        layer_file.holes = get_holes(&layer.path(), max_holes)?;
+                        n_deltas += 1;
+                    }
+                    layers.push(layer_file);
+                }
+            }
+            layers.sort_by_key(|layer| layer.lsn_range.end);
+
+            for i in 0..layers.len() {
+                if !layers[i].is_delta {
+                    let mut n_deltas_since_last_image = 0usize;
+                    let mut n_skipped = 0usize;
+                    let img_key_range = &layers[i].key_range;
+                    for j in (0..i).rev() {
+                        if range_overlaps(img_key_range, &layers[j].key_range) {
+                            if layers[j].is_delta {
+                                n_deltas_since_last_image += 1;
+                                if layers[j].skips(img_key_range) {
+                                    n_skipped += 1;
+                                }
+                            } else {
+                                break;
```

Comment:
This isn't technically correct, since this image might not cover the full key range. Maybe it's a good approximation but let's put a `// HACK` comment to make it clear.

There's another approximation: We should also check if this image is excess or not. If it's excess we can't break because the image layer wouldn't have existed if we used the new heuristic. **Your call whether it's worth fixing or just documenting.**

And there's yet another: The new layer map doesn't count total deltas, but the max stack of overlapping deltas.

Reply:
But probably long term we'd want the real layer_map+reimaging+compaction code to work only with metadata so that we can unit test it and use it here without rewriting it. It's not a small task though.

Reply:
Well, right now image layer is always dense, despite to the fact that it doesn't contain all possible key values in the specified range: there are may be no keys in the storage belonging to the image layer range but not present in the image layer. 

Comments added.